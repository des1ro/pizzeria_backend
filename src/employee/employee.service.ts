import { PizzeriaError } from "../exceptions/pizzeria.exceptions";
import { Employee } from "./employee";

export class EmployeeService {
  private readonly takenCheffs = new Set<Employee>();
  constructor(
    private readonly availableCheffs = new Set<Employee>(),
    private readonly availableWaiters = new Set<Employee>()
  ) {}
  getCheff() {
    const cheff = Array.from(this.availableCheffs).reverse().pop();
    if (cheff) {
      this.takenCheffs.add(cheff);
      return;
    }
    throw new PizzeriaError({
      name: "EMPLOYEE_ERROR",
      message: "There isn't free cheffs",
    });
  }
  releaseCheff(cheff: Employee) {
    if (this.takenCheffs.delete(cheff)) {
      this.availableCheffs.add(cheff);
      return;
    }
    throw new PizzeriaError({
      name: "EMPLOYEE_ERROR",
      message: "This cheff wasn't taken",
    });
  }
}
