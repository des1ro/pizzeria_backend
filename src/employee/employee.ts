export class Employee {
  constructor(readonly name: string, readonly role: EmployeeRole) {}
}
