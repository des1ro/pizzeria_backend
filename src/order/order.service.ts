import { PizzeriaError } from "../exceptions/pizzeria.exceptions";
import { OrderDTO } from "./order";

export class OrderService {
  constructor(
    private readonly completedOrders = new Set<OrderDTO>(),
    private readonly orderQueque = new Set<OrderDTO>(),
    private readonly orderInProgress = new Set<OrderDTO>()
  ) {}
  makeOrder() {
    //TODO
  }
  addOrderToQueque(order: OrderDTO) {
    if (this.orderQueque.has(order)) {
      throw new PizzeriaError({
        name: "ORDER_ERROR",
        message: "Order is already in queque",
      });
    }
    this.orderQueque.add(order);
  }
  makeOrderInProgress(order: OrderDTO) {
    if (this.orderQueque.delete(order)) {
      this.orderInProgress.add(order);
      return;
    }
    throw new PizzeriaError({
      name: "ORDER_ERROR",
      message: "There is no such as order in queque",
    });
  }
  compleateOrder(order: OrderDTO) {
    if (this.orderInProgress.delete(order)) {
      this.completedOrders.add(order);
      return;
    }
    throw new PizzeriaError({
      name: "ORDER_ERROR",
      message: "There is no order in order in progress",
    });
  }
  getCompletedOrders() {
    return this.completedOrders;
  }
}
