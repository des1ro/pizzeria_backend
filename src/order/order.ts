import { Ingredients } from "../pizza/ingredients.enum";
import { Pizza } from "../pizza/pizza";

export class OrderDTO {
  constructor(
    readonly id: number,
    readonly discount: Discount,
    readonly products: Pizza[],
    readonly tableId?: number
  ) {}
}
const pizza = new Pizza("anal", 20, [Ingredients.cheese]);
const order = new OrderDTO(1, Discount.codeDiscountOne, [pizza]);
