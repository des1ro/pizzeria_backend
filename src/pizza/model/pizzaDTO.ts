import { Ingredients } from "../enum/ingredients.enum";

export class PizzaDTO {
  constructor(
    readonly name: string,
    readonly price: number,
    readonly ingredients: Ingredients[]
  ) {}
}
