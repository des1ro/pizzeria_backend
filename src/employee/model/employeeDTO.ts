import { EmployeeRole } from "../enum/employee.enum";

export class EmployeeDTO {
  constructor(readonly name: string, readonly role: EmployeeRole) {}
}
