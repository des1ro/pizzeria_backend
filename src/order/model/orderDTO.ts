import { Discount } from "../../discounts/discount.enum";
import { PizzaDTO } from "../../pizza/model/pizzaDTO";

export class OrderDTO {
  constructor(
    readonly uuid: string,
    readonly discount: Discount,
    readonly products: PizzaDTO[],
    readonly tableId?: number
  ) {}
}
