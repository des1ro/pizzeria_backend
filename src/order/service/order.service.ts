import { EmployeeDTO } from "../../employee/model/employeeDTO";
import { PizzeriaError } from "../../exceptions/pizzeria.exceptions";
import { OrderDTO } from "../model/orderDTO";

export class OrderService {
  constructor(
    private readonly completedOrders = new Set<OrderDTO>(),
    private readonly orderQueque = new Set<OrderDTO>(),
    private readonly orderInProgress = new Map<OrderDTO, EmployeeDTO>()
  ) {}

  addOrderToQueque(order: OrderDTO): void {
    if (this.orderQueque.has(order)) {
      throw new PizzeriaError({
        name: "ORDER_ERROR",
        message: "Order is already in queque",
      });
    }
    this.orderQueque.add(order);
  }
  makeOrderInProgress(order: OrderDTO, cheff: EmployeeDTO) {
    if (this.orderQueque.delete(order)) {
      this.orderInProgress.set(order, cheff);
      return;
    }
    throw new PizzeriaError({
      name: "ORDER_ERROR",
      message: "There is no such as order in queque",
    });
  }
  completeOrderAndReturnCheff(order: OrderDTO): EmployeeDTO {
    const cheff = this.orderInProgress.get(order);
    if (this.orderInProgress.has(order) && cheff) {
      this.completedOrders.delete(order);
      this.completedOrders.add(order);
      return cheff;
    }
    throw new PizzeriaError({
      name: "ORDER_ERROR",
      message: "There is no order in order in progress",
    });
  }
  getOrderPrice(order: OrderDTO): number {
    return (
      (order.products.reduce((total, pizza) => total + pizza.price, 0) *
        (100 - order.discount)) /
      100
    );
  }
  getCompletedOrdersArray() {
    return Array.from(this.completedOrders);
  }
  getOrderId(): number {
    return (
      this.orderInProgress.size +
      this.orderQueque.size +
      this.completedOrders.size +
      1
    );
  }
}
