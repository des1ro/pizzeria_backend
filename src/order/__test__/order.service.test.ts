import { Discount } from "../../discounts/discount.enum";
import { EmployeeRole } from "../../employee/enum/employee.enum";
import { EmployeeDTO } from "../../employee/model/employeeDTO";
import { PizzeriaError } from "../../exceptions/pizzeria.exceptions";
import { PizzaDTO } from "../../pizza/model/pizzaDTO";
import { OrderDTO } from "../model/orderDTO";
import { OrderService } from "../service/order.service";
describe("Order service test suite", () => {
  let objectUnderTest: OrderService;
  let mockedOrderOne: OrderDTO;
  let mockedPizzas: PizzaDTO[];
  let mockedCheff: EmployeeDTO;
  beforeEach(() => {
    objectUnderTest = new OrderService();
  });
  beforeAll(() => {
    const mockedPizzaOne = new PizzaDTO("test pizza one", 20, []);
    const mockedPizzaTwo = new PizzaDTO("test pizza two", 50, []);
    mockedCheff = new EmployeeDTO("test employee", EmployeeRole.Cheff);
    mockedPizzas = [mockedPizzaOne, mockedPizzaTwo];
    mockedOrderOne = new OrderDTO(1, Discount.none, mockedPizzas);
  });
  it("Should add order to queque", () => {
    //When
    objectUnderTest.addOrderToQueque(mockedOrderOne);
    //Then
    expect(objectUnderTest).toMatchSnapshot();
  });
  it("Should throw PizzeriaError if order it's already in queque", () => {
    //Given
    //When
    objectUnderTest.addOrderToQueque(mockedOrderOne);
    //Then
    expect(() => objectUnderTest.addOrderToQueque(mockedOrderOne)).toThrow(
      PizzeriaError
    );
  });
  describe("Make order in progress test suite", () => {
    it("Should add order to make in progress", () => {
      //Given
      const mockedQueque = new Set<OrderDTO>();
      mockedQueque.add(mockedOrderOne);
      objectUnderTest = new OrderService(new Set<OrderDTO>(), mockedQueque);
      //When
      objectUnderTest.makeOrderInProgress(mockedOrderOne, mockedCheff);
      //Then
      expect(objectUnderTest).toMatchSnapshot();
    });
    it("Should throw PizzeriaError if order isn't in queque ", () => {
      //Then
      expect(() =>
        objectUnderTest.makeOrderInProgress(mockedOrderOne, mockedCheff)
      ).toThrow(PizzeriaError);
    });
  });
  describe("Complete order test suite", () => {
    it("Should complete order and return cheff", () => {
      //Given
      const mockedOrderInProgress = new Map<OrderDTO, EmployeeDTO>();
      mockedOrderInProgress.set(mockedOrderOne, mockedCheff);
      objectUnderTest = new OrderService(
        new Set<OrderDTO>(),
        new Set<OrderDTO>(),
        mockedOrderInProgress
      );
      //When
      const result =
        objectUnderTest.completeOrderAndReturnCheff(mockedOrderOne);
      //Then
      expect(objectUnderTest).toMatchSnapshot();
      expect(result).toBe(mockedCheff);
    });
    it("Should throw PizzeriaError if order wasn't in order in progress", () => {
      //Then
      expect(objectUnderTest).toMatchSnapshot();
      expect(() =>
        objectUnderTest.completeOrderAndReturnCheff(mockedOrderOne)
      ).toThrow(PizzeriaError);
    });
  });
  it("Should count order price and return value", () => {
    //Given
    const mockedOrderTwo = new OrderDTO(
      2,
      Discount.wednesdayForKidsDiscount,
      mockedPizzas
    );
    const expectedResultTwo = (20 + 50) * 0.9;
    //When
    const resultOne = objectUnderTest.getOrderPrice(mockedOrderOne);
    const expectedResultOne = 20 + 50;
    const resultTwo = objectUnderTest.getOrderPrice(mockedOrderTwo);
    //Then
    expect(resultOne).toBe(expectedResultOne);
    expect(resultTwo).toBe(expectedResultTwo);
  });
  it("Should get array of completed orders", () => {
    //Given
    const mockedOrderTwo = new OrderDTO(
      2,
      Discount.studentThursdayDiscount,
      []
    );
    const mockedCompletedOrders = new Set<OrderDTO>();
    mockedCompletedOrders.add(mockedOrderOne).add(mockedOrderTwo);
    //When
    objectUnderTest = new OrderService(mockedCompletedOrders);
    const resutlt = objectUnderTest.getCompletedOrdersArray();
    //Then
    expect(objectUnderTest).toMatchSnapshot();
    expect(resutlt).toBeInstanceOf(Array);
    expect(resutlt.length).toBe(2);
  });
  it("Should make new id for order", () => {
    //Given
    const mockedOrderTwo = new OrderDTO(2, Discount.codeDiscountOne, []);
    const mockedOrderThree = new OrderDTO(3, Discount.codeDiscountOne, []);
    const mockedOrderFour = new OrderDTO(4, Discount.codeDiscountOne, []);
    const mockedCompletedOrders = new Set<OrderDTO>();
    const mockedOrderQueque = new Set<OrderDTO>();
    const mockedOrderInProgress = new Map<OrderDTO, EmployeeDTO>();
    mockedCompletedOrders.add(mockedOrderOne).add(mockedOrderTwo);
    mockedCompletedOrders.add(mockedOrderThree);
    mockedOrderInProgress.set(mockedOrderFour, mockedCheff);
    objectUnderTest = new OrderService(
      mockedCompletedOrders,
      mockedOrderQueque,
      mockedOrderInProgress
    );
    //When
    const resultOne = objectUnderTest.getOrderId();
    const expectedResult = 2 + 1 + 1 + 1;
    //Then
    expect(resultOne).toBe(expectedResult);
    expect(objectUnderTest).toMatchSnapshot();
  });
});
