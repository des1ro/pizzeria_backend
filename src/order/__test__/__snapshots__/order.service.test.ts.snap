// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Order service test suite Complete order test suite Should complete order and return cheff 1`] = `
OrderService {
  "completedOrders": Set {
    OrderDTO {
      "discount": 0,
      "id": 1,
      "products": [
        PizzaDTO {
          "ingredients": [],
          "name": "test pizza one",
          "price": 20,
        },
        PizzaDTO {
          "ingredients": [],
          "name": "test pizza two",
          "price": 50,
        },
      ],
      "tableId": undefined,
    },
  },
  "orderInProgress": Map {
    OrderDTO {
      "discount": 0,
      "id": 1,
      "products": [
        PizzaDTO {
          "ingredients": [],
          "name": "test pizza one",
          "price": 20,
        },
        PizzaDTO {
          "ingredients": [],
          "name": "test pizza two",
          "price": 50,
        },
      ],
      "tableId": undefined,
    } => EmployeeDTO {
      "name": "test employee",
      "role": "Cheff",
    },
  },
  "orderQueque": Set {},
}
`;

exports[`Order service test suite Complete order test suite Should throw PizzeriaError if order wasn't in order in progress 1`] = `
OrderService {
  "completedOrders": Set {},
  "orderInProgress": Map {},
  "orderQueque": Set {},
}
`;

exports[`Order service test suite Make order in progress test suite Should add order to make in progress 1`] = `
OrderService {
  "completedOrders": Set {},
  "orderInProgress": Map {
    OrderDTO {
      "discount": 0,
      "id": 1,
      "products": [
        PizzaDTO {
          "ingredients": [],
          "name": "test pizza one",
          "price": 20,
        },
        PizzaDTO {
          "ingredients": [],
          "name": "test pizza two",
          "price": 50,
        },
      ],
      "tableId": undefined,
    } => EmployeeDTO {
      "name": "test employee",
      "role": "Cheff",
    },
  },
  "orderQueque": Set {},
}
`;

exports[`Order service test suite Should add order to queque 1`] = `
OrderService {
  "completedOrders": Set {},
  "orderInProgress": Map {},
  "orderQueque": Set {
    OrderDTO {
      "discount": 0,
      "id": 1,
      "products": [
        PizzaDTO {
          "ingredients": [],
          "name": "test pizza one",
          "price": 20,
        },
        PizzaDTO {
          "ingredients": [],
          "name": "test pizza two",
          "price": 50,
        },
      ],
      "tableId": undefined,
    },
  },
}
`;

exports[`Order service test suite Should get array of completed orders 1`] = `
OrderService {
  "completedOrders": Set {
    OrderDTO {
      "discount": 0,
      "id": 1,
      "products": [
        PizzaDTO {
          "ingredients": [],
          "name": "test pizza one",
          "price": 20,
        },
        PizzaDTO {
          "ingredients": [],
          "name": "test pizza two",
          "price": 50,
        },
      ],
      "tableId": undefined,
    },
    OrderDTO {
      "discount": 40,
      "id": 2,
      "products": [],
      "tableId": undefined,
    },
  },
  "orderInProgress": Map {},
  "orderQueque": Set {},
}
`;

exports[`Order service test suite Should make new id for order 1`] = `
OrderService {
  "completedOrders": Set {
    OrderDTO {
      "discount": 0,
      "id": 1,
      "products": [
        PizzaDTO {
          "ingredients": [],
          "name": "test pizza one",
          "price": 20,
        },
        PizzaDTO {
          "ingredients": [],
          "name": "test pizza two",
          "price": 50,
        },
      ],
      "tableId": undefined,
    },
    OrderDTO {
      "discount": 20,
      "id": 2,
      "products": [],
      "tableId": undefined,
    },
    OrderDTO {
      "discount": 20,
      "id": 3,
      "products": [],
      "tableId": undefined,
    },
  },
  "orderInProgress": Map {
    OrderDTO {
      "discount": 20,
      "id": 4,
      "products": [],
      "tableId": undefined,
    } => EmployeeDTO {
      "name": "test employee",
      "role": "Cheff",
    },
  },
  "orderQueque": Set {},
}
`;
